### PR TITLE
[FLINK-5185] [Table API & SQL] Decouple BatchTableSourceScan with TableSourceTable

### PR DESCRIPTION
**Pull request by beyond1920**
The pr aims to decouple BatchTableSourceScan with TableSourceTable for further optimization, e.g, projection pushdown/filter pushdown/query pushdown.  The principle is  let BatchTableSourceScan directly holding TableSource instead of holding TableSourceTable.
The main changes including:
1. change constructors of BatchScan/BatchTableSourceScan/DataSetScan
2. Extract the logic logical of build row types based on fields types and fields name into FlinkTypeFactory class.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchTableSourceScan.scala [hunk 1]
```diff
@@ -32,18 +31,20 @@ class BatchTableSourceScan(
     cluster: RelOptCluster,
     traitSet: RelTraitSet,
     table: RelOptTable,
-    rowType: RelDataType)
-  extends BatchScan(cluster, traitSet, table, rowType) {
+    val tableSource: BatchTableSource[_])
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchTableSourceScan.scala]: Do we have to declare this field as `val` ? It seems that it is only used in this class.

Comment (beyond1920) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchTableSourceScan.scala]: @wuchong , Thanks for your review. For later use, the projection pushdown rule needs to get the actual tablesource of the BatchTableSourceScan, then do something on it.

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchTableSourceScan.scala]: Get you !

**Comment by fhueske** [comment]
Thanks for the PR @beyond1920!
Good refactoring. 
+1 to merge

**Comment by fhueske** [comment]
merging

**Pull request closed without merging**